core/templates/heap.h

```diff
@@ -0,0 +1,155 @@
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef HEAP_H
+#define HEAP_H
+
+#include "core/error/error_macros.h"
+#include "core/typedefs.h"
+
+template <typename T>
+struct NoopIndexer {
+	void operator()(const T &p_value, uint32_t p_index) {}
+};
+
+/**
+ * A max-heap implementation that notifies of element index changes.
+ */
+template <typename T, typename LessThan = Comparator<T>, typename Indexer = NoopIndexer<T>>
+class Heap {
+	LocalVector<T> _buffer;
+
+	LessThan _less_than;
+	Indexer _indexer;
+
+public:
+	void reserve(uint32_t p_size) {
+		_buffer.reserve(p_size);
+	}
+
+	uint32_t size() const {
+		return _buffer.size();
+	}
+
+	void push(const T &p_element) {
+		_buffer.push_back(p_element);
+		_indexer(p_element, _buffer.size() - 1);
+		_shift_up(_buffer.size() - 1);
+	}
+
+	T pop() {
+		ERR_FAIL_COND_V_MSG(_buffer.is_empty(), T(), "Can't pop an empty heap");
+		T value = _buffer[0];
+		_indexer(value, UINT32_MAX);
+		if (_buffer.size() > 1) {
+			_buffer[0] = _buffer[_buffer.size() - 1];
+			_indexer(_buffer[0], 0);
+			_buffer.remove_at(_buffer.size() - 1);
+			_shift_down(0);
+		} else {
+			_buffer.remove_at(_buffer.size() - 1);
+		}
+		return value;
+	}
+
+	/**
+	 * Update the position of the element in the heap if necessary.
+	 */
+	void shift(uint32_t p_index) {
+		ERR_FAIL_UNSIGNED_INDEX_MSG(p_index, _buffer.size(), "Heap element index is out of range");
+		_shift_up(p_index) || _shift_down(p_index);
```

Comment:
```suggestion
		if (!_shift_up(p_index)) {
			_shift_down(p_index);
		}
```
Or something similar, the original syntax has very poor readability IMO